### ISSUE TITLE
Unmatched method on mocked object returns empty IEnumerable

### ISSUE THREAD

**Issue by rlenders** [moq/moq4#330]
When using Moq, calling an unmocked method that returns an `IEnumerable` on a mocked object returns an **empty** enumerable. I would have expected it to return `null`. If I change the method's signature to return `ICollection` or `IReadOnlyCollection`, it no longer returns an empty collection, but `null` as I would expect.

I have encountered numerous tests that rely on unmocked methods returning an empty enumerable. Those tests now fail because the methods were refactored to return `IReadOnlyCollection`.

This seems to be an inconsistency in the Moq framework.

**Comment by stakx** [comment]
> When using Moq, calling an unmocked method that returns an `IEnumerable` on a mocked object returns an empty enumerable. I would have expected it to return `null`.

I suspect that this is unlikely to change. Over time, special treatment for certain types has been built into the component of Moq that produces return values; that probably can't be reversed without breaking tests for other Moq users.

When you set `mock.DefaultValue = DefaultValue.Empty` (which is the default), the "empty default value provider" will currently treat the following return types in a special way:
 * array types of any rank
 * `IEnumerable` and `IEnumerable<T>`
 * `IQueryable` and `IQueryable<T>`
 * `Task` and `Task<T>`

> This seems to be an inconsistency in the Moq framework.

This is correct. As far as I know, at least the following inconsistencies exist:

* `IEnumerable` and `IEnumerable<T>` get special treatment, but other collection types don't. For others, Moq won't produce an empty collection, but simply `null`.
* When setting `mock.DefaultValue == DefaultValue.Mock`, most of Moq's special handling of the above-mentioned types goes away (or rather, isn't done in the way how `DefaultValue.Mock` supposedly works).

I'll look into this.

**Comment by stakx** [comment]
I've been thinking about this a little, and I am not sure that these inconsistencies can be fixed without breaking tests for people.

I currently see three ways out of this dilemma:

1. Follow through with #366 (currently closed for maintenance), i.e. make it possible to use custom default value providers for a mock. Then we can add a new default value provider implementation to mock that acts like `DefaultValue.Empty`, but has the inconsistencies cleaned up. People can then choose whether they want to stay on the previous behavior by not changing anything, or they can opt in to the new, more consistent provider.

2. Make the default value provider (`DefaultValue.Empty`) configurable with respect to the special types it recognises. Its default configuration would result in the current behavior, but special handling could be switched on or off for certain types. (For instance, you could enable special handling for `string` so that empty strings get returned instead of `null`.)

3. A combination of both of the above.

Next week, I should be ready to reopen my above PR (#366) that will enable (1). I'll think some more on (2).

**Comment by stakx** [comment]
Closing this issue in favor of #173. The main takeaway here is that Moq is at current not very consistent when it comes to default value handling for certain types, and this should be improved.

**Comment by stakx** [comment]
> When using Moq, calling an unmocked method that returns an IEnumerable on a mocked object returns an empty enumerable. I would have expected it to return null.

I just merged #536, which, together with #533 adds a very easy way of setting up a default value provider that will not produce "empty" default values, but work more like C#'s `default(T)` value expressions:

```csharp
public interface IFoo
{
    IEnumerable SomeEnumerable { get; }
}

sealed class NullOrDefaultValueProvider : LookupOrFallbackDefaultValueProvider { }

var mock = new Mock<IFoo> { DefaultValueProvider = new NullOrDefaultValueProvider() };
Assert.Null(mock.Object.SomeEnumerable);
```

**Comment by stakx** [comment]
@bojingo: You can create your mocks from a `MockRepository` which has a custom default value provider configured.
